raiden/utils/migrations/v18_to_v19.py

```diff
@@ -0,0 +1,162 @@
+            elif 'block_number' in statechange_data:
+                block_number = int(statechange_data['block_number'])
+                block_hash = cache.get(block_number)
+            data['triggered_by_block_hash'] = block_hash
+
+            updated_events.append(EventRecord(
+                event_identifier=event.event_identifier,
+                state_change_identifier=event.state_change_identifier,
+                data=json.dumps(data),
+            ))
+
+        storage.update_events(updated_events)
+
+
+def _transform_snapshot(
+        raw_snapshot: Dict[Any, Any],
+        storage: SQLiteStorage,
+        cache: BlockHashCache,
+) -> str:
+    """Upgrades a single snapshot by adding the blockhash to it and to any pending transactions"""
+    snapshot = json.loads(raw_snapshot)
+    block_number = int(snapshot['block_number'])
+    snapshot['block_hash'] = cache.get(block_number)
+
+    pending_transactions = snapshot['pending_transactions']
+    new_pending_transactions = []
+    for transaction_data in pending_transactions:
+        if 'raiden.transfer.events.ContractSend' not in transaction_data['_type']:
+            new_pending_transactions.append(transaction_data)
+            continue
+
+        # For each pending transaction find the corresponding DB event record.
+        event_record = storage.get_latest_event_by_data_field(
+            filters={'_type': transaction_data['_type']},
+        )
+        if not event_record.data:
+            raise InvalidDBData(
+                'Error during v18 -> v19 upgrade. Could not find a database event '
+                'table entry for a pending transaction.',
+            )
+
+        event_record_data = json.loads(event_record.data)
+        transaction_data['triggered_by_block_hash'] = event_record_data['triggered_by_block_hash']
+        new_pending_transactions.append(transaction_data)
+
+    snapshot['pending_transactions'] = new_pending_transactions
+    return json.dumps(snapshot)
+
+
+def _transform_snapshots_for_blockhash(storage: SQLiteStorage, cache: BlockHashCache) -> None:
+    """Upgrades the snapshots by adding the blockhash to it and to any pending transactions"""
+    updated_snapshots_data = []
+    for snapshot in storage.get_snapshots():
+        new_snapshot = _transform_snapshot(
+            raw_snapshot=snapshot.data,
+            storage=storage,
+            cache=cache,
+        )
+        updated_snapshots_data.append((new_snapshot, snapshot.identifier))
+
+    storage.update_snapshots(updated_snapshots_data)
+
+
+def upgrade_v18_to_v19(
+        storage: SQLiteStorage,
+        old_version: int,
+        current_version: int,
+        web3: Web3,
+) -> int:
+    if old_version == SOURCE_VERSION:
+        cache = BlockHashCache(web3)
+        _add_blockhash_to_state_changes(storage, cache)
+        _add_blockhash_to_events(storage, cache)
+        # The snapshot transformation should come last
```

Comment:
why? I believe it's because the update of the transaction queue relies on the update of the events, if it is, could you please add it to the comment?

Reply:
yes that's it. I will add it.

Reply:
Done in https://github.com/raiden-network/raiden/pull/3550/commits/9ceb9483d0c01fa538c1cf1df05bab52d5d1d3cc